sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,304 @@ def transform(self, X):
+                             "but n_neighbors=%d."
+                             % (X.shape[0], self.n_neighbors))
+
+        # Instantiate NN object, get column means, and store in statistics_
+        neigh = NearestNeighbors(n_neighbors=self.n_neighbors,
+                                 metric=self.metric)
+        self._fitted_neighbors = neigh.fit(X)
+        self.statistics_ = X_col_means
+
+        return self
+
+    def _transform(self, X, n_neighbors_new):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        n_neighbors_new : int
+            Indicates whether to pass n_neighbors or n_neighbors+1 to
+            _tranform().
+            Calling transform() automatically sets this to self.n_neighbors
+            while fit_transform() sets it to self.n_neighbors + 1.
+        """
+        # Import(s) here to avoid circular import
+        from ..neighbors.base import _get_weights
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed.")
+
+        # Get fitted data and ensure correct dimension
+        fitted_X = self._fitted_neighbors._fit_X
+        if X.shape[1] != fitted_X.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed.")
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+        if not np.any(row_total_missing > 0):
+            return X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. The "
+                "missing features in these rows are imputed with column means."
+                    .format(self.row_max_missing*100))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = _get_mask(X, self.missing_values)
+            row_total_missing = mask.sum(axis=1)
+        row_has_missing = row_total_missing.astype(np.bool)
+
+        if np.any(row_has_missing):
+            neighbors = self._fitted_neighbors.kneighbors(
+                X[row_has_missing, :], n_neighbors=n_neighbors_new)
+
+            # Get row index, distance, and weights of donors
+            knn_distances, knn_row_index = neighbors
+            # Remove self from list of donors
```

Comment:
Why? Surely self can't donate, because the feature is missing.

Reply:
It has been a while since I wrote this, but as far as I can remember the issue was that when using weights="distances", having "self" available meant that its weight would be 1 and other neighbors would get 0 weight.

Reply:
Well the same would apply if there was a duplicate training point that was not at the same index, but still had nothing to donate (please add a test case). So this isn't the right fix, unfortunately. I think you'll need to do the weighting over the masked array or something nasty like that. Or perhaps make _get_weights use np.inf instead of 1 and then handle that (squeeze to {0, 1} if any infs are unmasked) once you know what's masked.